---
title: "HybPiper 2"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval = FALSE)
```

HybPiper uses clean reads to create per samples assemblies and to extract the targeted sequences.  
See the full [HybPiper documentation](https://github.com/mossmatters/HybPiper/wiki/Full-pipeline-parameters) for more details.

**EXAMPLE FILES TO PROVIDE**

- `input_fastq.txt`
- namelist_<analysis_ID>.txt
- table of list of the samples and paths (spreadsheet)

## Preparation
The following assumes that:  

- We defined a unique identifier `<analysis_ID>` for the analysis and created a corresponding subfolder in the `DATA` directory (`<base_directory>/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/DATA`; see [here](Notes_for_advanced_wf.md) for details on folder organization and analysis naming)
- The analysis will be run in a temporary directory whose path is `path_to_tmp`
- The output folder will be stored in `<base_directory>/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/JOBS_OUTPUTS/`

As input, HybPiper needs:

- the list of the samples to analyse (`namelist_<analysis_ID>.txt`).There must not be any duplicate.  
- the clean reads for each sample (R1 and R2 .fastq). These can be either copied manually, or you can create the following text file to batch copy the .fastq files:  
  - `input_fastq.txt` [**ref to file**]: contains the copy commands (`scp`) to transfer the clean .fastq R1 and R2 files from their storage directory to the working directory. Each line contains a `scp`command specific to a single sample: `scp <path_to_clean_reads_storage>/Sample1_CLEAN* .`
  - additionally you can use `files_renaming.txt` [**ref to file**] to batch rename the input .fastq files. Each line contains a `rename` command specific to a single sample: `rename -v 'Sample1_CLEAN' 'Sample1' *`  
Note that `input_fastq.txt` and `files_renaming.txt` can be easily created using a simple spreadsheet (see [**insert ref to spreadsheet**])
- the target file `targetfile.fasta`

We assume `namelist_<analysis_ID>.txt`, `input_fastq.txt`, and `files_renaming.txt` are stored in the `DATA/<analysis_ID>` directory.


### Define the paths and variables

- for the inputs:
```{bash}
analysis_ID="FMN_104811_P007"
path_to_dir_in="/blue/soltis/dagallierl/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/DATA";

path_to_ref="/blue/soltis/dagallierl/DATASETS/PHYLOGENOMICS/target_references/hybpiper_check_targetfile"
reference_fasta_file="PROBE_SET_CLEAN_v4_prot_fixed.FAA"
```

- for the outputs:
```{bash}
step_ID="hybpiper2_assemble_no_stiched"
path_to_dir_out="/blue/soltis/dagallierl/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/JOBS_OUTPUTS/$analysis_ID"_"$step_ID"_"$SLURM_JOB_ID/";

```


## Assemble
```{bash}
ssh dagallierl@hpg2.rc.ufl.edu
cd /blue/soltis/dagallierl/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/SCRIPTS_cluster
sbatch hybpiper2_assemble_FMN_104811_P004.sh
sbatch hybpiper2_assemble_no_stiched_FMN_104811_P004.sh
```
59506237 in 13:25:05 (3-06:36:05 CPU)
59615989 in 11:11:39 (2-14:17:43 CPU)

## Extract
```{bash}
ssh dagallierl@hpg2.rc.ufl.edu
cd /blue/soltis/dagallierl/DATA_ANALYSES/PHYLOGENY_RECONSTRUCTION/SCRIPTS_cluster
sbatch hybpiper2_extract_no_stitched_FMN_104811_P004.sh
```
60195440 in 2 mins
